psdisk/psdisk.c

```diff
@@ -0,0 +1,220 @@
+	printf("\n\n");
+}
+
+
+static ssize_t psdisk_writeData(unsigned int addr, const void *buff, size_t bufflen)
+{
+	/* TO DO: add option to move partition table at the end of an image ?? */
+	return fwrite(buff, sizeof(uint8_t), bufflen, psdisk_common.file);
+}
+
+
+static int psdisk_createpTable(void)
+{
+	psdisk_common.mem.write = psdisk_writeData;
+
+	/* TODO: call ptable_writePartitions function */
+
+
+	return 0;
+}
+
+
+static int psdisk_parsePartitionOpt(int optind, char *argv[])
+{
+	int i;
+	char *name;
+	ptable_partition_t pHeader;
+
+	const int MAX_PARTITION_ATTR = 4;
+
+	for (i = 0; i < MAX_PARTITION_ATTR; ++i) {
+		if (argv[optind + i] == NULL) {
+			LOG_ERROR("Invalid number of arguments.");
+			return -1;
+		}
+	}
+
+	name = argv[optind];
+	for (i = 0; i < strlen(name); ++i) {
+		if (!isalnum(name[i])) {
+			LOG_ERROR("Invalid partition name - %s.", name);
+			return -1;
+		}
+	}
+
+	if (strlen(name) > sizeof(pHeader.name)) {
+		LOG_ERROR("Invalid partition name - %s.", name);
+		return -1;
+	}
+	else {
+		strncpy((char *)pHeader.name, name, strlen(name));
+	}
+
+	if ((pHeader.offset = strtoul(argv[optind + 1], NULL, 10)) == 0) {
+		LOG_ERROR("Partition %s has invalid offset. %d", pHeader.name);
+		return -1;
+	}
+
+	if ((pHeader.size = strtoul(argv[optind + 2], NULL, 10)) == 0) {
+		LOG_ERROR("Partition %s has invalid size.", pHeader.name);
+		return -1;
+	}
+
+	if ((pHeader.type = strtoul(argv[optind + 3], NULL, 16)) == 0) {
+		LOG_ERROR("Partition %s has invalid type.", pHeader.name);
+		return -1;
+	}
+
+	/* TODO: add pHeader to partitions list */
+
+	return 0;
+}
+
+
+static int psdisk_parseMemOpt(int optind, char *argv[])
+{
+	if (argv[optind] == NULL || argv[optind + 1] == NULL) {
+		LOG_ERROR("Invalid number of arguments.");
+		return -1;
+	}
+
+	if ((psdisk_common.mem.memSize = strtoul(argv[optind], NULL, 10)) == 0) {
+		LOG_ERROR("Invalid memory size.");
+		return -1;
+	}
+
+	if ((psdisk_common.mem.sectorSize = strtoul(argv[optind + 1], NULL, 10)) == 0) {
+		LOG_ERROR("Invalid sector size.");
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int psdisk_createImg(int optind, char *argv[])
+{
+	int i;
+	char *name;
+
+	if ((name = argv[optind]) == NULL) {
+		LOG_ERROR("Invalid number of arguments.");
+		return -1;
+	}
+
+	for (i = 0; i < strlen(name); ++i) {
+		if (!isalnum(name[i]) && name[i] != '/') {
+			LOG_ERROR("Invalid storage name - %s.", name);
+			return -1;
+		}
+	}
+
+	/* Open/create file in binary mode */
+	if ((psdisk_common.file = fopen(name, "w+b")) == NULL) {
+		LOG_ERROR("Cannot create image file, check path - %s.", name);
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static void psdisk_destroy(void)
+{
+	if (psdisk_common.file != NULL)
+		fclose(psdisk_common.file);
+}
+
+
+int main(int argc, char *argv[])
+{
+	int opt;
+
+	while ((opt = getopt(argc, argv, "i:s:p:h")) != -1) {
+		switch (opt) {
+			case 'i':
+				if (psdisk_createImg(optind - 1, argv) < 0)
```

Comment:
This is not how `getopt` should be used. By default `getopt` will thin any non-opt and non-optarg is positional argument and will permute `argv` to have all non-option arguments at the end. Example:
```
./pstable  -i asdf -p asdf 10 20 0x51 -i asdf

ARGV after getopt: ./pstable -i asdf -p asdf -i asdf 10 20 0x51
```
The least You can do is increment `optind` to skip additional parameters (so probably checking/extracting correct numer of parameters here is a good idea).
Or You can change how parameters are provided, ex. `-p asdf,10,20,0x51` which will make `getopt` behave correctly (but You will need to `strtok` the args).

Reply:
Also - You can provide `-i img` multiple times which will result in resource leak and strange results :).